apv5sdk-v15/apps/int6000-utils-1.1.6/int6k/int6krule.7
.TH int6krule 7 "Intellon Corporation" "int6000-utils-linux" "Intellon Linux Toolkit"
.SH NAME
int6krule - INT6x00 Stream Classification Utility
.SH SYNOPSIS
.BR int6krule
.RI [ options ] 
.IR action    
.IR operand 
.IR condition
.RI [ condition ]
.RI [ condition ]
.IR control
.IR volatility
.RI [ device ]
.RI [ device ]
[...]
.PP
where \fIcondition\fR is a \fIfield\fR \fIoperator\fR \fIvalue\fR sequence.
.SH DESCRIPTION
Format and send stream classification rules to one or more devices. Rules specify an action to be taken when a frame satisfies selection criteria. Selection criteria consists of one, two or three conditions where any or all conditions must be satisfied. Each condition consists of a field type, a relational operator and a value. Rules may be added to a device, or removed from a device, so that they have permanent or temporary effect.
.PP
Classification rules are cumulative. If a new rule set is identical to an old rule set then an error will occur unless it contains a different Transmission Action. In that case the old rule will be replaced. Identical classification rule sets are permitted if one of the sets is associated with a VLAN tag action. Classification is based on the original frame before is is altered by VLAN Tag insertion or removal. 
.PP
Classification is multi-dimensional and the terminology used here may seem strange at first. Refer to the Intellon Firmware Techncial Reference Manual description of the VS_CLASSIFICATION management message for a full explanation.
.PP
This program is part of the Intellon Linux Toolkit. See the \fBilt\fR man page for an overview and installation instructions.
.SH OPTIONS
.TP
.RB - e
Redirects stderr messages to stdout. By convention status and error messages are printed on stderr while primary program output is printed on stdout. This option prints all output on stdout in cases where this is desired.
.TP
.RB - q
Suppresses status messages on stderr. 
.TP
.RB - t
Print a list of program keywords on stdout. This option over-rides all others, except \fB-?\fR and \fB-!\fR, and the program will terminate without further action.
.TP
-\fBT \fItag\fR
The VLAN tag to be inserted into frames before they are transmitted. The tag is a 32-bit hexadecimal integer with optional "\fB0x\fR" prefix. This option is required for action \fBTagTX\fR and must be omitted for all other actions.  
.TP
.RB - v
Print additional information on stdout. In particular, this option dumps incoming and outgoing packets which can be saved as text files for reference.
.TP
-\fBV \fIversion\fR
The CSPEC version number expressed as a small decimal integer. This option is required (and should be \fB2\fR) for action \fBTagTX\fR and must be omitted for all other actions.  
.TP
.RB - ? ,-- help
Print program help summary on stdout. This option takes precedence over other options on the command line. 
.TP
.RB - ! ,-- version
Print program version information on stdout. This option takes precedence over other options on the command line. Use this option when sending screen dumps to Intellon Technical Support so that they know exactly which version of the Linux Toolkit you are using.
.SH ARGUMENTS
.TP
.IR action   
The action to be taken for frames that meet any (or all) selection criteria. Valid actions are listed and described under \fBACTIONS\fR.
.TP
.IR operand 
The operand specifies the logical relationship between conditions before the \fIaction\fR to be taken. Valid operands are listed and described under \fBOPERANDS\fR.
.TP
.IR condition
A conditional expression consisting of a \fIfield\fR, \fIoperator\fR and \fIvalue\fR. See \fBCONDITIONS\fR for more information.
.TP
.IR control
The control specifies the action to be taken by the device upon receipt of the rule. The basic actions are to add it to, or remove it from, the list of existing rules. Valid controls are listed and described under \fBCONTROLS\fR.
.TP
.IR volatility
The volatility specifies the effective lifetime of the rule. Temoprary rules are stored in SDRAM and are lost then the device is reset. Permanent rules are stored in NVRAM and are restored after the device is reset. Valid volatilities are listed and described under \fBVOLATILITY\fR.
.TP
.IR device
The MAC address of some powerline device. More than one address may be specified on the command line. If more than one address is specified then operations are performed on each device in turn. The default address is "\fBlocal\fR". See \fBDEVICES\fR for more information.
.SH ACTIONS
Actions indcate the disposition of frames that match selection criteria. They are expressed as discrete alphanumeric strings entered in upper, lower or mixed character case. They are position sensitive. Failure to enter a known action will results in an error message that lists all possible actions.
.TP
.BR CAP0 , CAP1 , CAP2 , CAP3
Assign a specific Channel Access Priority to frames.
.TP
.BR Drop , DropTX
Do not forward frames over powerline. 
.TP
.BR DropRX
Do not forward frames to host.
.TP
.BR Boost
Boost frame priority to CAP3 for MMEs only. At least one condition must be "\fBMME\fR". 
.TP
.BR StripTX
Remove the VLAN Tag from frames before transmission over powerline. This option checks for a VLAN Tag even when there are no VLAN related conditions.
.TP
.BR StripRX
Remove the VLAN Tag from frames before forwarding to host. This option checks for a VLAN Tag even when there are no VLAN related conditions.
.TP
.BR TagTX
Insert a given VLAN Tag into frames before transmission over powerline. This action requires option \fB-T\fR to specify the tag and option \fB-V\fR to specify the CSPEC version. 
.SH OPERANDS
The \fIoperand\fR indicates the logical relationship that must exist between conditions in the rule set before the action is applied to a frame. Operands are expressed as discrete alphanumeric strings entered in upper, lower or mixed character case. Failure to enter a known operand will result in an error message that lists all possible operands. They are positon sensitive. One operand is allowed and it must appear after the \fIaction\fR and before any \fIcondition\fR.
.TP
.BR Any
Apply the action to frames that satisfy \fBany\fR of the conditions. This is equivalent to the logical \fBor\fR operation.
.TP
.BR All
Apply the action to frames that satisfy \fBall\fR of the conditions. This is equivalent to the logical \fBand\fR operation.
.TP
.BR Always
Apply the action to all frames. All conditions are ignored.
.SH CONDITIONS
A condition consists of a \fIfield\fR, an \fIoperator\fR and a \fIvalue\fR. One condition is required but three are permitted. Condition order is not important but all conditions must appear after the \fIoperand\fR and before the \fIcontrol\fR. 
.TP
.IR field
The field is the part of the Ethernet frame to be examined. Some fields are not valid for some actions but this program does not enforce such rules since validation is performed by runtime firmware on each device. Recognized fields are listed and described under \fBFIELDS\fR. 
.TP
.IR operator
The operator specifies the relationsip that must exist between the \fIfield\fR and \fIvalue\fR in order for the \fIcondition\fR to evaluate \fBTrue\fR. Currently, only equality operators are supported. Valid operators are listed and described under \fBOPERATORS\fR.
.TP
.IR value
The value must be appropriate to the field type. Some fields are MAC or IP addresses, some are integers, some are bitmaps and others are states. Integers and bitmaps may be expressed in binary, decimal or hexadecimal format. Binary values staRt with \fB0b\fR. Hexadecimal values start with \fB0x\fR. States are expressed using keywords. Users are responsible for knowing how many bits are significant for each type of value.  Valid values are described along with fields under \fBFIELDS\fR. 
.SH FIELDS
Fields indicate the portion of the frame that is inspected during selection and the size and format of the value permited in the condition statement. They are expressed as discrete alphanumeric strings entered in upper, lower or mixed character case. Failure to enter a known field will result in an error message that lists all possible fields. 
.TP
.BR ET
A 16-bit Ethertype expressed in hexadecimal with optional "\fB0x\fR" prefeix. The format is described in IEEE Standard 802-2001 [4].
.TP 
.BR EthDA 
A 48-bit Ethernet destination address expressed in hexadecimal. Octets may be separated with optional colons for clarity. The format is described in IEEE Standard 802-2001 [4]. 
.TP 
.BR EthSA  
A 48-bit Ethernet source address expressed in hexadecimal. Octets may be separated with optional colons for clarity. The format is described in IEEE Standard 802-2001 [4]. 
.TP
.BR IPSP
A 16-bit IP source port expressed as a decimal integer. This condition applies to either TCP or UDP packets, depending on the protocol used, and is valid only for actions "\fBCAP0\fR", "\fBCAP1\fR", "\fBCAP2\fR", "\fBCAP3\fR" and "\fBDrop\fR". 
.TP
.BR IPDP 
A 16-bit IP destination port expressed as a decimal integer. This condition applies to either TCP or UDP packets, depending on the protocol used, and is valid only for actions "\fBCAP0\fR", "\fBCAP1\fR", "\fBCAP2\fR", "\fBCAP3\fR" and "\fBDrop\fR". 
.TP
.BR IPV4TOS 
An 8-bit Type-of-Service code where the format is defined in the RFC 791 (Internet Protocol) [14]. 
.TP
.BR IPV4PROT 
An 8-bit Ethernet protocol code. The format is defined in the RFC 791 (Internet Protocol) [14]. 
.TP
.BR IPV4SA
A 32-bit Internet Protocol source address expressed in dotted-decimal notation. The official format is defined in RFC 791 (Internet Protocol) [14]. Our implementation permits empty octets and leading zeros within fields. For example, "..." is equivalent to "0.0.0.0 and "127..000.001" is equivalent to "127.0.0.1". 
.TP
.BR IPV4DA 
A 32-bit Internet Protocol destination address expressed in dotted-decimal notation. The official format is defined in RFC 791 (internet Protocol) [14]. Our implementation permits empty octets and leading zeros within fields. For example, "..." is equivalent to "0.0.0.0 and "127..000.001" is equivalent to "127.0.0.1". 
.TP
.BR IPV6TC
An 8-bit Internet Protocol V6 traffic class expressed as defined in RFC 2460 (Internet Protocol Version 6) [17].
.TP
.BR IPV6FL
A 24-bit IPV6 flow label where the lower 20 bits are the IPv6 Flow Label defined in RFC 2460 (Internat Protocol Version 6) [17]. The upper 4 bits should be zero. The value can be entered either as a decimal, binary or hex integer.
.TP
.BR IPV6SA
A 128-bit IPV6 source address expressed as colon-separated hexadecmial quartets (octet pairs). The official format is defined in RFC 2460 (Internet Protocol Version 6) [17]. Our implementation permits multiple empty fields, abreviated fields and leading zeros within fields. When multiple empty fields appear, only the right-most occurance expands to zeros. For example, "FFFF::DDDD::BBBB::AAAA" is equivalent to "FFFF:0000:DDDD:0000:BBBB:0000:0000:AAAA".  
.TP
.BR IPV6DA
A 128-bit IPV6 destination address expressed as colon-separated hexadecimal quartets (octet pairs). The official format is defined in RFC 2460 (Internet Protocol Version 6) [17]. Our implementation permits multiple empty fields, abbreviated fields and leading zeros within fields. When multiple empty fields appear, only the right-most occurance expands to zeros. For example, "AAAA::BBBB::CCCC::DDDD" is equivalent to "AAAA:0000:BBBB:0000:CCCC:0000:0000:DDDD".  
.TP
.BR MME
A 24-bit Intellon HomePlugAV Management Message type expressed as a hex byte stream. For clarity, the recommeded format it "xx:xxxx". The first byte is the MMV. The next two bytes are the MMTYPE. Both are defined in the HomePlug AV Specification. The MMTYPE will match all MME variants, such as Request, Confirm, Indicate and Response because the lower two bits are ignored. This field is only valid for action "\fBBoost\fR". 
.TP
.BR TCPAck
The string "\fBTrue\fR" or "\fBFalse\fR" to indicate that the frame is (or is not) a TCP Acknowledgement. Double negatives are allowed so "Is True" is equvalent to "Not False" and "Is False" is equivalent to "Not True".
.TP
.BR TCPSP
A 16-bit TCP source port as a decimal integer. The format is defined in RFC 793 (Transmission Control Protocol [15]). 
.TP
.BR TCPDP 
A 16-bit TCP destination port expressed as a decimal integer. The format is defined in RFC 793 (Transmission Control Protocol [15]). 
.TP
.BR UDPSP 
A 16-bit UDP source port expressed as a decimal integer. The format is defined in RFC 768 (User Datagram Protocol [13]). 
.TP
.BR UDPDP
A 16-bit UDP destination port expressed as a decimal integer. The format is defined in RFC 768 (User Datagram Protocol [13]). 
.TP
.BR VLANID 
A 16-bit VLAN identifier where the lower 12 bits are the VLAN Identifier (VID) defined in IEEE Std 802.1Q-1998 (Virtual Bridged Local Area Networks) [11]. The upper 4 bits should be zero. 
.TP 
.BR VLANUP 
An 8-bit Ethernet VLAN tag where the lower 3 bits are the User Priority sub-field of a VLAN Tag defined in IEEE Std 802.1Q-1998 (Virtual Bridged Local Area Networks) [11]. The upper 5 bits should be zero. 
.TP
.BR VLANTag
The string "\fBPresent\fR" or "\fBMissing\fR" to indicate the presence (or absence) of one or more VLAN Tags within a frame. This classifier is essentially equivalent to "\fBET Is 0x8100\fR". Double negatives are allowed so "Is Present" is equivalent to "Not Missing" and "Is Missing" is equivalent to "Not Present".
.SH OPERATORS
An operator indicates an equality between a \fIfield\fR and a \fIvalue\fR. An operator is an alphanumeric string entered in upper, lower or mixed character case. Failure to enter a known operator will result in an error message that lists all possible operators. Operators are position sensitive and must appear between each \fIfield\fR and \fIvalue\fR.
.TP
.BR Is
Indicates that the frame field must equal the associated value for the condition to evaluate true.
.TP
.BR Not
Indicates that the frame field must not equal the associated value for the condition to true true.
.SH STATES
A state is a special case of \fIvalue\fR.                     
.TP
.BR True , On , Yes , Present
Indicates a positive state or presence of some entity. All are equivalent and can be used interchangeably. Double negatives are permitted so "Is True" is equvalent to "Not False".
.TP
.BR False , Off , No , Missing
Indicates a negative state or absence of some entity. ALl are equivalent and can be used interchangeably.  Double negatives are permitted so "Is False" is equvalent to "Not True".
.SH CONTROLS
The control determines how the devices will handle the rule after it is validated. It is expressed as a discrete alphanumeric string entered in upper, lower or mixed character case. Failure to enter a known control will result in an error message that lists all possible controls. The control is position sensitive and must occur after \fIcondition\fR and before \fIvolatility\fR. 
.TP
.BR Add
Adds the rule to the current list of rules unless a violation occurs. In some cases, a rule may replace an existing rule instead of being added.
.TP
.BR Rem , Remove
Remove the rule from the current list of rules unless a violation occurs.
.SH VOLATILITY
The volatility determines which device rule set will be affected by the action. It is expressed as a discrete alphanumeric string entered in upper, lower or mixed character case. Failure to enter a known volatility will result in an error message that lists all possible volatilities. The volatility is position sensitive and must occur after \fIcontrol\fR. 
.TP
.BR Temp
The temporary rule set will be modified. The temporary rule set resides in the working PIB stored in SDRAM.   
.TP
.BR Perm
The permanent rule set will be modified. The permanent rule set resides in the user PIB stored in NVRAM.
.SH DEVICES
Powerline devices use Ethernet Media Access Control (MAC) addresses. A MAC address is a 48-bit value entered as 12 hexadecimal digits in upper, lower or mixed character case. Octets may be separated with colons for clarity. For example, "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are valid and equivalent.
.PP
The following MAC addresses are special and may be entered by name instead of number.
.TP
.BR all
Same as "broadcast".
.TP
.BR broadcast
A synonym for the Ethernet broadcast address, \fBFF:FF:FF:FF:FF:FF\fR. All devices, whether local, remote or foreign recognize messages sent to this address.  A remote device is any device at the far end of a powerline connection. A foreign device is any device not manufactured by Intellon.
.TP
.BR local
A synonym for the Intellon vendor specific Local Management Address (LMA), \fB00:B0:52:00:00:01\fR. All local Intellon devices recognize this address but remote and foreign devices do not. A remote device is any device at the far end of a powerline connection. A foreign device is any device not manufactured by Intellon.
.SH REFERENCES
See the Intellon HomePlug AV Firmware Technical Reference Manual for more information.
.SH DISCLAIMER
Intellon HomePlug AV Vendor Specific Management Message structure and content is proprietary to Intellon Corporation, Ocala FL USA. Consequently, public information is not available. Intellon Corporation reserves the right to modify message structure or content in future firmware releases without any obligation to notify or compensate users of this program.
.SH EXAMPLES
This command adds a temporary classification rule to the classification table on device B00:B0:52:BA:BE:01. The rule instructs the device to drop frames that match either (any) of two conditions. The first condition states that the IPv4 source address is 192.168.99.2. The second conditon states that the IPv4 destination address is 192.168.99.100.  
.PP
   # int6krule drop any IPv4SA is 192.168.99.2 IPv4DA is 192.168.99.100 add temp 00:B0:52:BA:BE:01
.PP
Observe that the \fIaction\fR, \fIoperand\fR and \fIconditions\fR come first then the \fIcontrol\fR and \fIvolatility\fR then the affected \fIdevices\fR. Up to three conditions may be specified. Keyword order is important. Character case is not important. 
.PP
The following example prints a list of programmed keywords on stdout for reference. The example shown here has been abbreviate due to formatting limitations.
.PP
   # int6krule -t
    
     Controls are 'Add'|'Rem'|'Remove'
     Volatilities are 'Temp'|'Perm'
     Actions are 'CAP0'|'CAP1'|'CAP2'|'CAP3'|'Boost'|...|'StripTX'|'StripRX'|'TagRX'
     Operands are 'All'|'Any'|'Always'
     Fields are 'EthDA'|'EthSA'|'VLANUP'|'VLANID'|'IPv4TOS'|...|'TCPAck'|'VLANTag'
     Operators are 'Is'|'Not'

.PP
More example follow:
.PP    
.B  Ethernet Address Rules
.PP
Ethernet address rules have the following general format:
.PP 
   | CAP0 | ANY | EthSA | IS  | xx:xx:xx:xx:xx:xx | ADD    | TEMP | xx:xx:xx:xx:xx:xx
   | CAP1 | ALL | EthDA | NOT |                   | REMOVE | PERM |
   | CAP2 |
   | CAP3 |
   | DROP |
.PP 
For example, instruct device 00:B0:52:BA:BE:FF to temporarily add a rule to forward frames from 00:2B:FE:CA:FE:BA at CAP3. Observe Ethernet hardware addresses are used both in the condition and for the affected powerline devices.
.PP 
   # int6krule cap3 any ethsa is 00:2b:fe:ca:fe:ba add temp 00:b0;52:ba:be:ff
.PP
.B IP Address Rules
.PP
IP address rules have the following general format:
.PP 
   | CAP0 | ANY | IPv4SA | IS  | ddd.ddd.ddd.ddd | ADD    | TEMP | xx:xx:xx:xx:xx:xx
   | CAP1 | ALL | IPv4DA | NOT |                 | REMOVE | PERM |
   | CAP2 |
   | CAP3 |
   | DROP |
 
.PP
For example, instruct device 00:B0:52:BA:BE:FF to permanently remove the rule that drops packets from 192.168.99.1. Notice that the IP address is specified in dotted decimal format but the device address is specified in hexadecimal octet format. Dotted decimal format permits empty and variable length octets but octet delimitors are mandatory. Hexadecimal octet format requires fixed length octets but octet delimiters are optional.
.PP 
   # int6krule drop any ipv4sa is 192.168.99.1 remove perm 00:b0:52:ba:be:ff
.PP
.B IP Protocol Rules
.PP
IP protocol rules have the following general format:
.PP 
   | CAP0 | ANY | IPv4PROT | IS  | xxxx | ADD    | TEMP | xx:xx:xx:xx:xx:xx
   | CAP1 | ALL |          | NOT |      | REMOVE | PERM |
   | CAP2 |
   | CAP3 |
   | DROP |
.PP 
For example, to instruct device 00:B0:52:BA:BE:FF to permanently add a rule to forward non-IP packets at CAP2. In this example, delmiters have been omitted from the device Ethernet address.
.PP 
   # int6krule CAP2 all ipv4prot not 0x0800 add perm 00b052babeff
.SH SEE ALSO
.BR ilt ( 7 ),
.BR int6krate ( 7 ),
.BR int6kstat ( 7 ),
.BR int6ktone ( 7)
.SH CREDITS
 Charles Maier <charles.maier@intellon.com>


apv5sdk-v15/apps/int6000-utils-1.1.6/int6k/int6krule.c
/*====================================================================*
 *   
 *   Copyright (c) 2010, Atheros Communications Inc.
 *   
 *   Permission to use, copy, modify, and/or distribute this software 
 *   for any purpose with or without fee is hereby granted, provided 
 *   that the above copyright notice and this permission notice appear 
 *   in all copies.
 *   
 *   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL 
 *   WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED 
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL  
 *   THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR 
 *   CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM 
 *   LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, 
 *   NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN 
 *   CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *   
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   int6krule.c - Intellon Stream Classification Utility;
 *
 *   this program sends and receives raw ethernet frames and so needs
 *   root priviledges; if you install it using "chmod 555" and "chown 
 *   root:root" then you must login as root to run it; otherwise, you 
 *   can install it using "chmod 4555" and "chown root:root" so that 
 *   anyone can run it; the program will refuse to run until you get
 *   things right;
 *   
 *   This software and documentation is the property of Intellon 
 *   Corporation, Ocala, Florida. It is provided 'as is' without 
 *   expressed or implied warranty of any kind to anyone for any 
 *   reason. Intellon assumes no responsibility or liability for 
 *   errors or omissions in the software or documentation and 
 *   reserves the right to make changes without notification. 
 *   
 *   Intellon customers may modify and distribute the software 
 *   without obligation to Intellon. Since use of this software 
 *   is optional, users shall bear sole responsibility and 
 *   liability for any consequences of it's use. 
 *
 *.  Intellon INT6x00 Linux Toolkit for HomePlug AV; 
 *:  Published 2007 by Intellon Corp. ALL RIGHTS RESERVED;
 *;  For demonstration; Not for production use;
 *
 *   Contributor(s):
 *      Charles Maier <charles.maier@intellon.com>
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*"
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../int6k/int6k.h"
#include "../int6k/rules.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/basespec.c"
#include "../tools/hexdump.c"
#include "../tools/hexencode.c"
#include "../tools/bytespec.c"
#include "../tools/ipv4spec.c"
#include "../tools/ipv6spec.c"
#include "../tools/hexdecode.c"
#include "../tools/todigit.c"
#include "../tools/endian.c"
#include "../tools/error.c"
#include "../tools/synonym.c"
#include "../tools/assist.c"
#include "../tools/lookup.c"
#include "../tools/expect.c"
#endif

#ifndef MAKEFILE
#include "../int6k/Confirm.c"
#include "../int6k/Display.c"
#include "../int6k/Failure.c"
#include "../int6k/ReadMME.c"
#include "../int6k/Request.c"
#include "../int6k/SendMME.c"
#include "../int6k/devices.c"
#include "../int6k/EthernetHeader.c"
#include "../int6k/IntellonHeader.c"
#include "../int6k/UnwantedPacket.c"
#include "../int6k/rules.c"
#include "../int6k/ParseRule.c"
#endif

#ifndef MAKEFILE
#include "../ether/channel.c"
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#endif

#ifndef MAKEFILE
#include "../mme/MMECode.c"
#endif

/*====================================================================*
 *
 *   signed Classification (struct int6k * int6k, struct rule * rule);
 *
 *   int6k.h
 *
 *   This plugin for program int6krule adds or removes a temporary
 *   or permanent network classification rule to a device using a
 *   VS_CLASSIFICATION message;
 *
 *   See the Intellon HomePlug AV Firmware Technical Reference Manual 
 *   for more information;
 *   
 *   This software and documentation is the property of Intellon 
 *   Corporation, Ocala, Florida. It is provided 'as is' without 
 *   expressed or implied warranty of any kind to anyone for any 
 *   reason. Intellon assumes no responsibility or liability for 
 *   errors or omissions in the software or documentation and 
 *   reserves the right to make changes without notification. 
 *   
 *   Intellon customers may modify and distribute the software 
 *   without obligation to Intellon. Since use of this software 
 *   is optional, users shall bear sole responsibility and 
 *   liability for any consequences of it's use. 
 *   
 *.  Intellon INT6x00 Linux Toolkit for HomePlug AV; 
 *:  Published 2007 by Intellon Corp. ALL RIGHTS RESERVED;
 *;  For demonstration; Not for production use;
 *
 *   Contributor(s):
 *      Charles Maier <charles.maier@intellon.com>
 *
 *--------------------------------------------------------------------*/

signed Classification (struct int6k * int6k, struct rule * rule) 

{

	struct channel * channel = (struct channel *)(int6k->channel);
	struct message * message = (struct message *)(int6k->message);
#ifndef __GNUC__
#pragma pack (push,1)
#endif

	struct __packed vs_classification_request 
	{
		struct header_eth ethernet;
		struct header_int intellon;
		struct rule rule;
	}
	* request = (struct vs_classification_request *)(message);
	struct __packed vs_classification_confirm 
	{
		struct header_eth ethernet;
		struct header_int intellon;
		uint8_t MSTATUS;
	}
	* confirm = (struct vs_classification_confirm *)(message);

#ifndef __GNUC__
#pragma pack (pop)
#endif

	Request (int6k, "Set Classification Rules");
	memset (message, 0, sizeof (struct message));
	EthernetHeader (&message->ethernet, channel->peer, channel->host);
	IntellonHeader (&message->intellon, (VS_CLASSIFICATION | MMTYPE_REQ));
	int6k->packetsize = sizeof (struct vs_classification_request);
	memcpy (&request->rule, rule, sizeof (request->rule));
	if (SendMME (int6k) <= 0) 
	{
		error ((int6k->flags & INT6K_BAILOUT), errno, INT6K_CANTSEND);
		return (-1);
	}
	if (ReadMME (int6k, (VS_CLASSIFICATION | MMTYPE_CNF)) <= 0) 
	{
		error ((int6k->flags & INT6K_BAILOUT), errno, INT6K_CANTREAD);
		return (-1);
	}
	if (confirm->MSTATUS) 
	{
		Failure (int6k, INT6K_WONTDOIT);
		return (-1);
	}
	Confirm (int6k, "Setting ...");
	return (0);
}

/*====================================================================*
 *   
 *   int main (int argc, const char * argv[]);
 *   
 *.  Intellon INT6x00 Linux Toolkit for HomePlug AV; 
 *:  Published 2007 by Intellon Corp. ALL RIGHTS RESERVED;
 *;  For demonstration; Not for production use;
 *
 *--------------------------------------------------------------------*/

int main (int argc, const char * argv []) 

{
	extern struct channel channel;
	extern const struct _term_ devices [];
	static const char *optv [] = 
	{
		"ei:qtT:vV:",
		"action operand condition [...] control volatility [device] [...]\n\n          where a condition is: field operator value",
		"Intellon Stream Classification Utility",
		"e\tredirect stderr messages to stdout",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface number [2]",

#else

		"i s\thost interface name [" CHANNEL_ETHDEVICE "]",

#endif

		"q\tquiet mode",
		"t\tdisplay symbol tables",
		"T x\tinserted vlan tag is x [0x00000000]",
		"v\tverbose mode",
		"V n\tcspec version is n [0]",
		(const char *) (0) 
	};

#include "../int6k/int6k-struct.c"

	struct cspec cspec;
	struct rule rule;
	signed c;
	memset (&rule, 0, sizeof (rule));
	memset (&cspec, 0, sizeof (cspec));
	if (getenv (PLCDEVICE)) 
	{
#if defined (WINPCAP) || defined (LIBPCAP)
		channel.id = atoi (getenv (PLCDEVICE));
#else
		channel.name = strdup (getenv (PLCDEVICE));
#endif
	}
	optind = 1;
	opterr = 1;
	while ((c = getoptv (argc, argv, optv)) != -1) 
	{
		switch ((char) (c)) 
		{
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.id = atoi (optarg);

#else

			channel.name = optarg;

#endif

			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			_setbits (int6k.flags, INT6K_SILENCE);
			break;
		case 't':
			printf ("\n");
			printf (" Controls are ");
			expect (controls, CONTROLS, stdout);
			printf ("\n");
			printf (" Volatilities are ");
			expect (volatilities, VOLATILITIES, stdout);
			printf ("\n");
			printf (" Actions are ");
			expect (actions, ACTIONS, stdout);
			printf ("\n");
			printf (" Operands are ");
			expect (operands, OPERANDS, stdout);
			printf ("\n");
			printf (" Fields are ");
			expect (fields, FIELDS, stdout);
			printf ("\n");
			printf (" Operators are ");
			expect (operators, OPERATORS, stdout);
			printf ("\n");
			printf (" States  are ");
			expect (states, STATES, stdout);
			printf ("\n");
			printf ("\n");
			return (0);
		case 'T':
			cspec.VLAN_TAG = (uint32_t)(basespec (optarg, 16, sizeof (cspec.VLAN_TAG)));
			cspec.VLAN_TAG = htonl (cspec.VLAN_TAG);
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (int6k.flags, INT6K_VERBOSE);
			break;
		case 'V':
			cspec.CSPEC_VERSION = (uint16_t)(basespec (optarg, 10, sizeof (cspec.CSPEC_VERSION)));
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (ParseRule (&argc, &argv, &rule, &cspec) == -1) 
	{
		error (1, 0, "invalid rule");
	}

#ifndef WIN32

/*
 *      raw packets require root priviledges on linux; one does not have to be
 *      root when this program is installed setuid using 'chown root:root' and 
 *      'chmod 4555';
 */

	if (geteuid ()) 
	{
		error (1, EPERM, INT6K_NEEDROOT);
	}

#endif

	openchannel (&channel);
	if (!(int6k.message = malloc (sizeof (struct message)))) 
	{
		error (1, errno, INT6K_ERR_MEMORY);
	}
	if (!argc) 
	{
		Classification (&int6k, &rule);
	}
	while ((argc) && (*argv)) 
	{
		if (!hexencode (channel.peer, sizeof (channel.peer), synonym (*argv, devices, DEVICES))) 
		{
			error (1, errno, INT6K_BAD_MAC, *argv);
		}
		Classification (&int6k, &rule);
		argc--;
		argv++;
	}
	free (int6k.message);
	closechannel (&channel);
	exit (0);
}

